language/tools/move-bytecode-utils/src/layout.rs

```diff
@@ -158,55 +174,115 @@ impl<'a, T: GetModule> SerdeLayoutBuilder<'a, T> {
+                panic!(
+                    "Could not find struct named {} in module {}",
+                    name,
+                    declaring_module.borrow().name()
+                )
+            });
+        let normalized_struct = Struct::new(declaring_module.borrow(), def).1;
+        assert_eq!(
+            normalized_struct.type_parameters.len(),
+            type_arguments.len(),
+            "Wrong number of type arguments for struct"
+        );
+
+        let generics: Vec<String> = type_arguments
+            .iter()
+            .zip(normalized_struct.type_parameters.iter())
+            .filter_map(|(type_arg, type_param)| {
+                if self.config.ignore_phantom_types && type_param.is_phantom {
+                    // do not include phantom type arguments in the struct key, since they do not affect the struct layout
+                    None
+                } else {
+                    Some(print_format_type(type_arg))
+                }
+            })
+            .collect();
+        let struct_key_module = if self.config.omit_addresses {
+            format!("{}", module_id.name())
+        } else {
+            format!(
+                "{}{}{}",
+                module_id.address(),
+                self.config.separator.as_deref().unwrap_or("::"),
+                module_id.name()
+            )
+        };
+        let struct_key = if generics.is_empty() {
+            format!(
+                "{}{}{}",
+                struct_key_module,
+                self.config.separator.as_deref().unwrap_or("::"),
+                name
+            )
         } else {
-            let generics: Vec<String> = type_arguments.iter().map(print_format_type).collect();
-            format!("{}::{}<{}>", module_id, name, generics.join(","))
+            format!(
+                "{}{}{}{}{}{}",
+                struct_key_module,
+                self.config.separator.as_deref().unwrap_or("::"),
+                name,
+                self.config.separator.as_deref().unwrap_or("<"),
+                generics.join(self.config.separator.as_deref().unwrap_or(",")),
+                self.config.separator.as_deref().unwrap_or(">")
+            )
         };
-        if !shallow && !self.registry.contains_key(&struct_key) {
-            let declaring_module = self
-                .module_resolver
-                .get_module_by_id(module_id)?
-                .expect("Failed to resolve module");
-            let def = declaring_module
-                .borrow()
-                .find_struct_def_by_name(name)
-                .unwrap_or_else(|| {
+        if self.config.shallow {
+            return Ok(Format::TypeName(struct_key));
+        }
+
+        if let Some(old_struct) = self.registry.get(&struct_key) {
+            if self.config.omit_addresses {
+                // check for conflicts (e.g., 0x1::M::T and 0x2::M::T that both get stripped to M::T because
+                // omit_addresses is on)
```

Comment:
Ah, I think this is what I was talking about in my top level comment. Should we report conflicts regardless?